core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        _c_l - count given label (count of each label)
+        """
+        # Ensure that results are deterministic
+        # There is nothing special about the value 4
+        numpy.random.seed(seed=4)
+
+        self._alpha = self._DEFAULT_ALPHA
+        self._beta = self._DEFAULT_BETA
+
+    def _get_word_id(self, word):
+        """Returns a word's id if it exists, otherwise assigns
+        a new id to the word and returns it."""
+        try:
+            return self._word_to_id[word]
+        except KeyError as ke:
+            self._word_to_id[word] = self._word_count
+            self._word_count += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """Returns a word's id if it exists, otherwise assigns
+        a new id to the label and returns it."""
+        try:
+            return self._label_to_id[label]
+        except KeyError as ke:
+            self._label_to_id[label] = self._label_count
+            self._label_count += 1
+        return self._label_to_id[label]
+
+    def _get_record(self, d):
+        """Given a doc id, return the doc and labels."""
+        return self._w_dc[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """Returns a vector specifying which labels
+        are specified in the training set."""
+        if len(labels) == 0:
+            return numpy.ones(self._label_count)
+        label_vector = numpy.zeros(self._label_count)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        # Always set default label
+        label_vector[self._label_to_id['_default']] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, val):
+        """Update counting matrices given doc/word/label indices."""
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._doc_count)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dc.append(l_c)
+            for w, l in zip(doc, l_c):
+                self._update_counting_matrices(d, w, l, 1)
+
+    def _infer_docs(self, doc_ids):
+        """Runs iterative inference on given docs (default all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for c in xrange(len(doc)):
+                w = doc[c]
+                l = self._l_dc[d][c]
```

Comment:
The tricky part about l is to figure out whether it represents a label string, a label id, or a boolean. Perhaps we should rename it to label_id?